docs/manual/buildsystem.md

```diff
@@ -0,0 +1,115 @@
+---
+layout: default
+title: rpm.org - Declarative builds
+---
+
+# Declarative builds
+
+Most software follows common patterns as to how it should be prepared and
+built, such as the classic `./configure && make && make install` case.
+In a spec, these all go to their own sections and when combined with
+the steps to unpack sources, it creates a lot of boilerplate that is
+practically identical across a lot of specs. With the odd tweak to
+this or that detail for distro preferences and the like. To reduce
+the repetitive boilerplate, rpm >= 4.20 adds support for a declarative
+build system mechanism where these common steps can be defined centrally
+per each build system. Packagers then only need to declare which build
+system they use, and optionally supply additional switches and/or
+arguments for the steps where needed.
+
+## Spec usage
+
+In the most simple case, a spec will have an buildsystem declaration such
+as `BuildSystem: cmake` and no manually specified build scripts at all.
+However it's common to have to manually tweak a thing or two. There are
+several ways how to accomplish this, what's appropriate depends on the
+case.
+
+1) Use append and/or prepend as necessary. Need to delete a file after
+the install step? Add a `%install -a` section with the appropriate `rm`.
+Need to `sed` something just before the build step? Add a `%build -p`
+section with the necessary steps.
+
+2) Another very common need is to pass extra arguments to the build
+commands, build configuration in particular. This is done with the
+BuildOption tag, which can appear arbitrary number of times
+in the spec for each section.
+
+```
+BuildOption(<section>): <option string>
+```
+
+without the parenthesis defaults to configuration. In other words,
+these two lines are exactly equivalent:
+
+```
+BuildOption: --enable-fu
+BuildOption(conf): --enable-fu
+```
+
+Passing these per-section options to the actual buildsystem of the
+package is the responsibility of the buildsystem specific macros.
+
+3) Complex packages can have things like multiple build systems, in
+which case you might want to invoke the macros manually, eg.
+
+```
+%buildsystem_autotools_build
```

Comment:
This is probably the least favorite part of mine, in terms of verbosity, but I suppose it can be tweaked later. Also, it's not the common happy path anyway so, as a packager needing an advanced/non-standard build procedure, once you open the lid and peek inside, you're already on your own anyway :smile:

Reply:
To elabore a bit (since I was quite unspecific above), what I mean by "verbose" is that this macro name consists of 3 separate parts, each of which I'm surely going to need to look up every time I use them (which, again, shouldn't be too often) :smile:

Reply:
Oh! I like that. Good idea.

Reply:
Just spent a good part of the morning on that alias thing again, only to realize that it's largely useless because BuildOption's are not accessible from the spec, so manually invoking these macros in their long or the alleged shorter form is kinda pointless. Oh well.